File: 2023/src/day2.zig
const std = @import("std");
const ascii = @import("std").ascii;

const fileData = @embedFile("./files/day2.txt");
const fileDataEx = @embedFile("./files/day2ex.txt");
//const cregex = @import("ctregex");
const Regex = @import("zig-regex").Regex;

const cubes = struct { r: i32, g: i32, b: i32 };

fn get_match(line: []const u8, color: []const u8) !i32 {
    var arena_state = std.heap.ArenaAllocator.init(std.heap.c_allocator);
    defer arena_state.deinit();
    const allocator = arena_state.allocator();
    var re = try Regex.compile(allocator, color);

    var captures = try re.captures(line);
    var result: i32 = 0;
    if (captures != null) {
        var cap = captures.?.sliceAt(1);
        result = try std.fmt.parseInt(i32, cap.?, 10);
    }
    return result;
}

fn get_full_match(line: []const u8) !cubes {
    var result: cubes = undefined;
    result.r = try get_match(line, "(\\d+) red");
    result.b = try get_match(line, "(\\d+) blue");
    result.g = try get_match(line, "(\\d+) green");
    return result;
}

fn solve(data: anytype, rule: cubes) !void {
    var sum_part1: i32 = 0;
    var sum_part2: i32 = 0;

    var splits = std.mem.split(u8, data, "\n");
    var game: i32 = 1;
    while (splits.next()) |line| {
        var subsplit = std.mem.split(u8, line, ";");
        var valid_game: bool = true;
        var maxs = cubes{ .r = 0, .g = 0, .b = 0 };

        // Treat one game
        while (subsplit.next()) |item| {
            var res = try get_full_match(item);
            if (res.r > rule.r or res.g > rule.g or res.b > rule.b) {
                valid_game = false;
            }
            maxs.r = @max(res.r, maxs.r);
            maxs.g = @max(res.g, maxs.g);
            maxs.b = @max(res.b, maxs.b);
        }

        if (valid_game) {
            //std.debug.print("{s} -> Valid game {}\n", .{ line, game });
            sum_part1 += game;
        }
        //std.debug.print("Game {} ->  {any} = {}\n", .{ game, maxs, (maxs.r * maxs.g * maxs.b) });
        sum_part2 += (maxs.r * maxs.g * maxs.b);
        game += 1;
    }

    std.debug.print("Part 1 = {}, Part 2 = {}\n", .{ sum_part1, sum_part2 });
}

pub fn main() !void {
    const argv = std.os.argv;
    var example: bool = false;
    std.log.info("Hello day 3: {s}", .{argv});
    for (argv) |args| {
        if (std.mem.eql(u8, args[0..2], "ex")) {
            example = true;
        }
    }

    std.debug.print("Hello day 2!\n", .{});
    if (example) {
        try solve(fileDataEx, cubes{ .r = 12, .g = 13, .b = 14 });
    } else {
        try solve(fileData, cubes{ .r = 12, .g = 13, .b = 14 });
    }
}

File: 2023/src/day2.py

import regex as re
import numpy as np

# 12 red cubes, 13 green cubes, and 14 blue cubes
num_reg = re.compile("(\d+)")
regex_str = [re.compile(f"(\d+) {x}") for x in ["red", "green", "blue"]]
rules = [12,13,14]

def valid_record(record, rules=[]):
    for reg, rule in zip(regex_str, rules):
        a=reg.search(record)
        if a and int(a.group(1)) > rule:
            return False
        # elif rule > 0:
        #     return False
    return True

def get_rgb(record):
    res = []
    for reg, rule in zip(regex_str, rules):
        a=reg.search(record)
        val = int(a.group(1)) if a else 0
        res.append(val)
    return res
            

def main():
  
    #"(\d)+ red", "(\d)+ blue"]
    result = 0
    part2=0
    with open("./src/files/day2.txt", "r") as f:
        for line in f.readlines():
            #print(line.strip())
            game = line.strip().split(":")[0]
            #print(game)
            num_game = int(num_reg.search(game).group(1))
           
            if np.all([valid_record(record, rules= rules) for record in line.strip().split(":")[1].split(";")]):
                result += num_game
                print(f" {line.strip()} -> valid")
            
            rs=[]
            gs = []
            bs = []
            for record in line.strip().split(":")[1].split(";"):
                r = get_rgb(record)
                #print(r, record)
                rs.append(r[0])
                gs.append(r[1])
                bs.append(r[2])
            #print(max(rs),max(bs),max(gs))
            cube_power = max(rs) * max(bs) * max(gs)
            #print(cube_power)
            part2+=cube_power
            # min_b = max(bs)
            # min_g = max(gs)
            # for record in line.strip().split(":")[1].split(";"):
            #     if valid_record(record, rules= rules):
            #         print(f"  {record} -> valid")
                    
              

    print(f"Part 1 res = {result}")
    print(f"Part 2 res = {part2}")


if __name__ == "__main__":
    main()
